Mvom increment

philfuster opened the pull request:
### Summary

This PR implements the new `increment` operation.

#### TypeScript Changes

- Created Input and Output types for new `increment` operation
- Created async static method `increment` on dynamically generated Model class via `compileModel`.
- Created `transformPathToOrdinalPosition` to transform an increment operation input key path to its ordinal position.
- Updated `Connection` to handle new `RecordNotFoundError`.

#### UniBasic Changes

- Created unibasicTemplate - `increment.njk`
    - validates increment operation input
    - Has a read lock loop to handle locked records.
        -  sleeps according to the passed in retry delay.
        - returns `RecordLockedError` when retry max is exceeded
    - returns new `RecordNotFoundError` error when `else` statement executes on `readu`.
    -  writes record after performing incrementing.
    - returns the updated document
  
### Testing Notes

Ran tests using MVOM:

- incrementing a record that does not exist - received `RecordNotFoundError`
- incrementing one field - field was incremented, updated document returned
- incremented multiple fields - fields were incremented, updated document returned
- incremented nested field - nested field was incremented, updated document returned
- incremented value position - value position was incremented
- incremented multiple value positions - value positions were incremented
- incremented sub value positions - sub value positions were incremented
- incremented while record was locked - retries were performed - `RecordLockedError` returned.

shawnmcknight commented on src/Schema.ts:
```diff
@@ -277,6 +277,61 @@ class Schema<
 	}
 
+	/** Transform the paths to ordinal positions. Returning a '.' delimited string. Ex. "1.2.3"  */
+	public transformPathsToOrdinalPositions(paths: string[]): string[] {
```
This method feels weird for several reasons:
1. You are accepting a `string[]` and returning a `string[]` but the only use case calls for passing a single value in and getting a single value back.
2. The logic in the `else if` seems improper. For the projection use case, if a parent key was supplied then it needed to also produce paths for all children. In this use case, the expectation would be that a full keypath is supplied and not partial keypaths.
3. The `paths` isn't really a `string[]`. It's specifically the keys from the flattened document structure.
4. The return type is specifically `${number} | ${number}.${number} | ${number}.${number}.${number}`.
5. If the path isn't found then returning empty string or potentially an empty list seems improper. That would then set the `path` that gets passed to basic as an empty string or undefined value and then cause problems on the basic side.

philfuster replied:
> 3. The paths isn't really a string[]. It's specifically the keys from the flattened document structure.
I'm trying to get this to work, but I'm struggling with how can I pass the Schema to `FlattenDocument` or `InferSchemaPath` when I'm within Schema. I've tried passing `this` and `typeof this` as the type parameter, but that wasn't working.

shawnmcknight replied:
This worked for me:
```
	/** Transform the paths to ordinal positions. Returning a '.' delimited string. Ex. "1.2.3"  */
	public transformPathsToOrdinalPositions(
		paths: Extract<
			keyof FlattenDocument<typeof this, Schema | SchemaDefinition | SchemaTypeDefinitionNumber>,
			string
		>[],
	): string[] {
	```
Although I would recommend making an alias for that and using it in both places it would be needed.

philfuster replied:
I was getting infinite type instantiation as well

shawnmcknight replied:
```
	public transformPathsToOrdinalPositions(
		path: InferSchemaPaths<Schema<TSchemaDefinition>>,
	): string {
	```
	
	Since that method is on `Schema` it doesn't require the same restrictions to the paths that the increment method requires. So I think this will work.

philfuster replied:
That works but then the type of the paths is a union of  `string | number | symbol`. I'm casting the value to `String`, but I'm not 100% sure that's accurate. I'm not entirely sure why it's a union of the 3 types either.

philfuster replied:
Interesting I would not have thought to just extract the string type from that. Why does that work? It makes sense what Extract does, but why is it valid/the right thing to do, to just ignore the other types?

I wrap `InferSchemaPaths` as such: `Extract<InferSchemaPaths<Schema<TSchemaDefinition>>, string>` and that appears to work within `transformPathsToOrdinalPositions`, but I end up with `never` when using `transformPathsToOrdinalPositions`.
![image](https://github.com/user-attachments/assets/a24a6377-704b-4755-a557-194bcd8b6f53)

philfuster replied:
I'm thinking it has to do with the `TDictionaries` as it defaults to `Record<never, never>`

philfuster replied:
but passing `TDictionaries` along does not help.

shawnmcknight commented on src/Schema.ts:
```diff
@@ -277,6 +277,61 @@ class Schema<
+		return [...positions];
+	}
+
+	private buildOrdinalPathString(ordinalPath: number[]): string {
```
This feels like we're relying on a quirk of UniBasic where it will let you have empty strings for values and subvalues. It also means this code is more complex than it needs to be. It could simply be:

```
		const [attributePosition, valuePosition = 0, subvaluePosition = 0] = ordinalPath;
		return `${attributePosition + 1}.${valuePosition + 1}.${subvaluePosition + 1}`;
```

That code is then simple enough that it can probably just be consolidated into the `transformPathsToOrdinalPositions` method (especially since I think that can be slimmed down) and it won't need additional checks for length and such.

philfuster replied:
Consolidated into `transformPathsToOrdinalPositions`

shawnmcknight reviewed (changes requested):
I've finished a first pass review and left several comments. Overall this looks good, but I had suggestions for improvement.

philfuster commented on src/compileModel.ts:
```diff
@@ -51,6 +56,25 @@ export interface ModelFindByIdOptions extends ModelDatabaseExecutionOptions {
+	 */
+	retry?: number;
+	/**
+	 * between retries in seconds when record is locked
```
changed

philfuster commented on src/compileModel.ts:
```diff
@@ -51,6 +56,25 @@ export interface ModelFindByIdOptions extends ModelDatabaseExecutionOptions {
 }
+export interface ModelIncrementOptions extends ModelDatabaseExecutionOptions {
+	/**
+	 * number retries to perform when record is locked
```
done

philfuster commented on src/compileModel.ts:
```diff
@@ -288,6 +312,47 @@ const compileModel = <TSchema extends Schema | null>(
+			options: ModelIncrementOptions = {},
+		): Promise<ModelCompositeValue<TSchema>> {
+			const { maxReturnPayloadSize, requestId, userDefined, retry = 5, retryDelay = 1 } = options;
+			if (this.schema == null) {
```
makes sense. removed the check

philfuster commented on src/compileModel.ts:
```diff
@@ -326,6 +391,33 @@ const compileModel = <TSchema extends Schema | null>(
+			}
+
+			return operations.map(({ path, value }) => {
+				if (this.schema == null) {
```
good idea, thank you. done.

philfuster commented on src/unibasicTemplates/subroutines/external/increment.njk:
```diff
@@ -0,0 +1,185 @@
+
+* create null projection
+if udoCreate(UDO_NULL, projection) then
+   call error_handler(ERROR_UDO, output)
```
done. Fixed in save as well.

philfuster commented on src/unibasicTemplates/subroutines/external/increment.njk:
```diff
@@ -0,0 +1,185 @@
+    go closeReleaseAndReturnFromSub
+  end
+
+  * parse out path.
```
Good idea. Done.

philfuster commented on src/types/DbSubroutineOutput.ts:
```diff
@@ -31,6 +31,11 @@ export interface DbSubroutineOutputFindByIds {
 export type DbSubroutineResponseFindByIds = DbSubroutineResponse<DbSubroutineOutputFindByIds>;
 
+export interface DbSubroutineOutputIncrement {
+	result: DbDocument | null;
```
I don't think it is. I will remove `null` form the output type.

philfuster commented on src/compileModel.ts:
```diff
@@ -288,6 +312,47 @@ const compileModel = <TSchema extends Schema | null>(
+					...(userDefined && { userDefined }),
+				},
+			);
+			if (data.result == null) {
```
yea I don't think `null` is a valid response. I have removed this check.

philfuster commented on src/unibasicTemplates/subroutines/external/increment.njk:
```diff
@@ -0,0 +1,185 @@
+end
+
+* write out record
+write record on f.file, recordId on error
```
Created this jira task - https://storis.atlassian.net/browse/CXM-8262

philfuster commented:
> I've finished a first pass review and left several comments. Overall this looks good, but I had suggestions for improvement.

@shawnmcknight Thank you for your review. I have addressed your feedback in the latest series of commits. This is ready for another look.

shawnmcknight commented on src/unibasicTemplates/subroutines/external/increment.njk:
```diff
@@ -0,0 +1,189 @@
+
+  * parse out path.
+  pathMatchTest = path match "1n'.'1n'.'1n"
+  if pathMatchTest eq 0 then
```
This is returning a 1 or a 0 which is a boolean true/false in basic. It should be treated as boolean for clarity.

You can consolidate this into a single line expression:
`if not(path matches "1n'.'1n'.'1n") then`

philfuster replied:
Done thank you.

shawnmcknight commented on src/compileModel.ts:
```diff
@@ -51,6 +56,25 @@ export interface ModelFindByIdOptions extends ModelDatabaseExecutionOptions {
+
+export type IncrementOperation<TSchema extends Schema | null> = TSchema extends Schema
+	? {
+			path: keyof FlattenDocument<TSchema, Schema | SchemaDefinition | SchemaTypeDefinitionNumber>;
```
This should also have the `Extract` on it since the called method expects it.

philfuster replied:
Done. added Extract here.

shawnmcknight commented on src/__tests__/Schema.test.ts:
```diff
@@ -388,6 +388,67 @@ describe('transformPathsToDbPositions', () => {
+	} satisfies SchemaDefinition;
+	const schema = new Schema(definition);
+
+	test('should throw error if path list is empty', () => {
```
This test description doesn't seem to align with the test.

philfuster replied:
done. Thank you.

shawnmcknight reviewed (changes requested):
Looks good. Just a few minor comments left.

philfuster commented on src/Schema.ts:
```diff
@@ -278,6 +278,26 @@ class Schema<
+	): `${number}.${number}.${number}` {
+		const positionPaths = this.getPositionPaths();
+
+		if (!positionPaths.has(path)) {
```
Agreed. done.

philfuster commented on src/Schema.ts:
```diff
@@ -278,6 +278,26 @@ class Schema<
+	 * @throws {Error} Invalid path provided
+	 */
+	public transformPathToOrdinalPosition(
+		path: Extract<InferSchemaPaths<Schema<TSchemaDefinition, TDictionaries>>, string>,
```
Yea that makes a lot more sense. Done

philfuster replied:
Do you know how `number` and `symbol` end up in that resulting type?

shawnmcknight reviewed (approved):
LGTM!
